pkg/apiserver/registry/system/supportbundle/rest.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundle
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/afero"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog"
+	"k8s.io/utils/exec"
+
+	systemv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/system/v1beta1"
+	"github.com/vmware-tanzu/antrea/pkg/ovs/ovsctl"
+	"github.com/vmware-tanzu/antrea/pkg/support"
+)
+
+const bundleExpireDuration = time.Hour
+
+var (
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+)
+
+// NewStorage creates a support bundle storage. The mode should be either agent
+// or controller. If the mode is agent, the client argument should not be nil.
+func NewStorage(mode string, client ovsctl.OVSCtlClient) Storage {
+	bundle := &supportBundleREST{
+		mode:         mode,
+		ovsCtlClient: client,
+		cache: &supportBundle{
+			SupportBundle: systemv1beta1.SupportBundle{
+				ObjectMeta: metav1.ObjectMeta{Name: mode},
+				Status:     systemv1beta1.SupportBundleStatusNone,
+			},
+		},
+	}
+	return Storage{
+		Mode:          mode,
+		SupportBundle: bundle,
+		Download:      &downloadREST{supportBundle: bundle},
+	}
+}
+
+// Storage contains REST resources for support bundle, including status query and download.
+type Storage struct {
+	SupportBundle *supportBundleREST
+	Download      *downloadREST
+	Mode          string
+}
+
+var (
+	_ rest.Scoper          = &supportBundleREST{}
+	_ rest.Getter          = &supportBundleREST{}
+	_ rest.Creater         = &supportBundleREST{}
+	_ rest.GracefulDeleter = &supportBundleREST{}
+)
+
+type supportBundle struct {
+	systemv1beta1.SupportBundle
+	filepath string
+}
```

Comment:
It looks like this struct is not really necessary as the file and the bundle object don't share lifecycle and not always created together.
Another thought is, maybe we could just put filepath into the SupportBundle struct since it already has its metadata in the struct, no harm to let the API expose the local filename too, right? Then I think some code could be cleaner in this file.

Reply:
Updated. The FilePath in the type struct is a hidden field.
FYI @jianjuns.

Reply:
I would not expose the internal FS path in API. If we are able to hide it I am fine.

Reply:
Then maybe keep it as is, I just felt a wrapper struct is not necessary when both of them can be accessed in supportBundleREST.